File: implement-shell-tools/ls/ls.js
const fs = require("fs");

const args = process.argv.slice(2);

let oneLine = false;
let showHidden = false;
let targets = [];

for (const arg of args) {
    if (arg === "-1") {
        oneLine = true;
    } else if (arg === "-a") {
        showHidden = true;
    } else {
        targets.push(arg);
    }
}

if (targets.length === 0) {
    targets.push(".");
}

function listDirectory(directory) {
    let files = fs.readdirSync(directory);

    if (!showHidden) {
        files = files.filter(file => !file.startsWith("."));
    }

    files.sort();

    if (oneLine) {
        console.log(files.join("\n"));
    } else {
        console.log(files.join("  "));
    }
}

function listTarget(target) {
    try {
        const stats = fs.statSync(target);

        if (stats.isDirectory()) {
            listDirectory(target);
        } else {
            console.log(target);
        }

    } catch (error) {
        console.error(`ls: cannot access '${target}': No such file or directory`);
    }
}

for (const target of targets) {
    listTarget(target);
}

File: implement-shell-tools/ls/ls.py
import sys
import os

args = sys.argv[1:]

one_line = False
show_hidden = False
targets = []

for arg in args:
    if arg == "-1":
        one_line = True
    elif arg == "-a":
        show_hidden = True
    else:
        targets.append(arg)

if len(targets) == 0:
    targets.append(".")


def list_directory(directory):
    files = os.listdir(directory)

    if not show_hidden:
        files = [file for file in files if not file.startswith(".")]

    files.sort()

    if one_line:
        print("\n".join(files))
    else:
        print("  ".join(files))


def list_target(target):
    try:
        if os.path.isdir(target):
            list_directory(target)
        else:
            print(target)

    except Exception:
        print(f"ls: cannot access '{target}': No such file or directory")


for target in targets:
    list_target(target)